### PR TITLE
mgr: synchronize ClusterState's health and mon_status.

### PR DESCRIPTION
**Pull request by rzarzynski**
Fixes: https://tracker.ceph.com/issues/24995
Signed-off-by: Radoslaw Zarzynski <rzarzyns@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/mgr/ClusterState.h [hunk 1]
```diff
@@ -139,6 +136,24 @@ class ClusterState
       pg_map,
       std::forward<Args>(args)...);
   }
+
+  template<typename Callback, typename...Args>
+  auto with_health(Callback&& cb, Args&&...args) const ->
+    decltype(cb(health_json, std::forward<Args>(args)...))
```
Comment (rzarzynski) [inline src/mgr/ClusterState.h]: SIMPLIFY ME.

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2020-03-29_08:38:08-rados-wip-kefu-testing-2020-03-29-1503-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-03-29_14:23:20-rados-wip-kefu-testing-2020-03-29-1503-distro-basic-smithi

**Comment by rzarzynski** [comment]
@tserong: please take a look on [the gist](https://gist.github.com/rzarzynski/cb9c1d1efe99d003c52ec88deecc4123#conclusion).

**Comment by tserong** [comment]
Excellent writeup @rzarzynski, thank you!

**Pull request merged**